src/common/cockpitconf.c

```diff
@@ -21,92 +21,152 @@
 
 #include "cockpitconf.h"
 
-#include "cockpithash.h"
-
-#include "errno.h"
-
-static GHashTable *cockpit_conf = NULL;
-static GHashTable *cached_strvs = NULL;
-
-const gchar *cockpit_config_file = "cockpit.conf";
-const gchar *cockpit_config_dirs[] = { PACKAGE_SYSCONF_DIR, NULL };
-
-static gboolean
-load_key_file (const gchar *file_path)
+#include <assert.h>
+#include <ctype.h>
+#include <err.h>
+#include <errno.h>
+#include <libgen.h>
+#include <limits.h>
+#include <regex.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+/* #define DEBUG 1 */
+#if DEBUG
+#define debug(fmt, ...) (fprintf (stderr, "cockpitconf: " fmt "\n", ##__VA_ARGS__))
+#else
+#define debug(...)
+#endif
+
+
+typedef struct Entry {
+    char *section;
+    char *key;
+    char *value;
+    char **strv_cache;
+    char strv_delimiter;
+    struct Entry *next;
+} Entry;
+
+static bool cockpit_conf_loaded = false;
+static Entry *cockpit_conf = NULL;
+
+const char *cockpit_config_file = "cockpit.conf";
+const char *cockpit_config_dirs[] = { PACKAGE_SYSCONF_DIR, NULL };
+
+static char*
+strlower (char *s)
 {
-  GKeyFile *key_file = NULL;
-  GHashTable *section;
-  GError *error = NULL;
-  gchar **groups;
-  gint i;
-  gint j;
-  gboolean ret = TRUE;
+  for (char *c = s; *c; ++c)
+    *c = tolower (*c);
+  return s;
+}
 
-  key_file = g_key_file_new ();
+/* this is like reallocarray(3), but this does not yet exist everywhere */
+static void*
+reallocarr (void *ptr, size_t nmemb, size_t size)
+{
+  if (nmemb >= SIZE_MAX / size)
+    err (1, "reallocarr: overflow (nmemb %zu)", nmemb);
+  return realloc (ptr, nmemb * size);
+}
 
-  if (!g_key_file_load_from_file (key_file, file_path, G_KEY_FILE_NONE, &error))
+/* See https://developer.gnome.org/glib/stable/glib-Key-value-file-parser.html for the spec,
+ * except that cockpit.conf is case insensitive for section and key names; so store them in lower case */
+static bool
+load_key_file (const char *file_path)
+{
+  FILE *f = NULL;
+  char *cur_section = NULL;
+  regex_t re_section, re_keyval, re_ignore;
+  char *line = NULL;
+  bool ret = true;
+  size_t line_size = 0;
+  ssize_t line_len;
+
+  cockpit_conf_loaded = true;
+
+  f = fopen (file_path, "r");
+  if (!f)
     {
-      if (g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_NOENT))
-        goto out;
-
-      g_message ("couldn't load configuration file: %s: %s", file_path, error->message);
-      ret = FALSE;
-      goto out;
+      if (errno != ENOENT)
+        warnx ("couldn't load configuration file: %s: %m\n", file_path);
+      return false;
     }
 
-  groups = g_key_file_get_groups (key_file, NULL);
-  for (i = 0; groups[i] != NULL; i++)
+  assert (regcomp (&re_section, "^[[:space:]]*\\[ *([^][:space:]]+) *\\][[:space:]]*$", REG_EXTENDED|REG_NEWLINE) == 0);
+  assert (regcomp (&re_keyval, "^[[:space:]]*([^=[:space:]]+)[[:space:]]*=[[:space:]]*(.*)$", REG_EXTENDED|REG_NEWLINE) == 0);
```

Comment:
key names are `[:alnum:]` plus dash (`-`)

Reply:
Is `[[:space:]]*(.*)` guaranteed not to capture space?  ie: do we have a promise of first-encountered-is-most-greedy?

Reply:
Yes, regex(7) specifies that. However, this is also moot with the stricter key format,  then we don't need to rely on greediness any more.